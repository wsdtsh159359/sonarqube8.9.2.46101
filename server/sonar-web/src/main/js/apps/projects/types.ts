/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { ComponentQualifier } from '../../types/component';

export interface Project {
  analysisDate?: string;
  isFavorite?: boolean;
  key: string;
  leakPeriodDate?: string;
  measures: T.Dict<string>;
  name: string;
  projects?: number;
  qualifier: ComponentQualifier;
  tags: string[];
  visibility: T.Visibility;
  needIssueSync?: boolean;
}

export interface Facet {
  [value: string]: number;
}

export interface Facets {
  [property: string]: Facet;
}
