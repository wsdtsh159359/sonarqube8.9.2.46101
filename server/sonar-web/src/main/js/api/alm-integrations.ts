/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { get, getJSON, parseError, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  AzureProject,
  AzureRepository,
  BitbucketProject,
  BitbucketRepository,
  GithubOrganization,
  GithubRepository,
  GitlabProject
} from '../types/alm-integration';
import { ProjectBase } from './components';

export function setAlmPersonalAccessToken(almSetting: string, pat: string): Promise<void> {
  return post('/api/alm_integrations/set_pat', { almSetting, pat }).catch(throwGlobalError);
}

export function checkPersonalAccessTokenIsValid(
  almSetting: string
): Promise<{ status: boolean; error?: string }> {
  return get('/api/alm_integrations/check_pat', { almSetting })
    .then(() => ({ status: true }))
    .catch(async (response: Response) => {
      if (response.status === 400) {
        const error = await parseError(response);
        return { status: false, error };
      }
      return throwGlobalError(response);
    });
}

export function getAzureProjects(almSetting: string): Promise<{ projects: AzureProject[] }> {
  return getJSON('/api/alm_integrations/list_azure_projects', { almSetting }).catch(
    throwGlobalError
  );
}

export function getAzureRepositories(
  almSetting: string,
  projectName: string
): Promise<{ repositories: AzureRepository[] }> {
  return getJSON('/api/alm_integrations/search_azure_repos', { almSetting, projectName }).catch(
    throwGlobalError
  );
}

export function searchAzureRepositories(
  almSetting: string,
  searchQuery: string
): Promise<{ repositories: AzureRepository[] }> {
  return getJSON('/api/alm_integrations/search_azure_repos', { almSetting, searchQuery }).catch(
    throwGlobalError
  );
}

export function importAzureRepository(
  almSetting: string,
  projectName: string,
  repositoryName: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_azure_project', {
    almSetting,
    projectName,
    repositoryName
  }).catch(throwGlobalError);
}

export function getBitbucketServerProjects(
  almSetting: string
): Promise<{ projects: BitbucketProject[] }> {
  return getJSON('/api/alm_integrations/list_bitbucketserver_projects', { almSetting });
}

export function getBitbucketServerRepositories(
  almSetting: string,
  projectName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    projectName
  });
}

export function importBitbucketServerProject(
  almSetting: string,
  projectKey: string,
  repositorySlug: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_bitbucketserver_project', {
    almSetting,
    projectKey,
    repositorySlug
  }).catch(throwGlobalError);
}

export function searchForBitbucketServerRepositories(
  almSetting: string,
  repositoryName: string
): Promise<{
  isLastPage: boolean;
  repositories: BitbucketRepository[];
}> {
  return getJSON('/api/alm_integrations/search_bitbucketserver_repos', {
    almSetting,
    repositoryName
  });
}

export function getGithubClientId(almSetting: string): Promise<{ clientId?: string }> {
  return getJSON('/api/alm_integrations/get_github_client_id', { almSetting });
}

export function importGithubRepository(
  almSetting: string,
  organization: string,
  repositoryKey: string
): Promise<{ project: ProjectBase }> {
  return postJSON('/api/alm_integrations/import_github_project', {
    almSetting,
    organization,
    repositoryKey
  }).catch(throwGlobalError);
}

export function getGithubOrganizations(
  almSetting: string,
  token: string
): Promise<{ organizations: GithubOrganization[] }> {
  return getJSON('/api/alm_integrations/list_github_organizations', {
    almSetting,
    token
  }).catch((response?: Response) => {
    if (response && response.status !== 400) {
      throwGlobalError(response);
    }
  });
}

export function getGithubRepositories(data: {
  almSetting: string;
  organization: string;
  pageSize: number;
  page?: number;
  query?: string;
}): Promise<{ repositories: GithubRepository[]; paging: T.Paging }> {
  const { almSetting, organization, pageSize, page = 1, query } = data;
  return getJSON('/api/alm_integrations/list_github_repositories', {
    almSetting,
    organization,
    p: page,
    ps: pageSize,
    q: query || undefined
  }).catch(throwGlobalError);
}

export function getGitlabProjects(data: {
  almSetting: string;
  page?: number;
  pageSize?: number;
  query?: string;
}): Promise<{ projects: GitlabProject[]; projectsPaging: T.Paging }> {
  const { almSetting, pageSize, page, query } = data;
  return getJSON('/api/alm_integrations/search_gitlab_repos', {
    almSetting,
    projectName: query || undefined,
    p: page,
    ps: pageSize
  })
    .then(({ repositories, paging }) => ({ projects: repositories, projectsPaging: paging }))
    .catch(throwGlobalError);
}

export function importGitlabProject(data: {
  almSetting: string;
  gitlabProjectId: string;
}): Promise<{ project: ProjectBase }> {
  const { almSetting, gitlabProjectId } = data;
  return postJSON('/api/alm_integrations/import_gitlab_project', {
    almSetting,
    gitlabProjectId
  }).catch(throwGlobalError);
}
