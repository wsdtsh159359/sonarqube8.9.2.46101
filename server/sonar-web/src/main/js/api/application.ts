/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { getJSON, post, postJSON } from 'sonar-ui-common/helpers/request';
import throwGlobalError from '../app/utils/throwGlobalError';
import { Application, ApplicationPeriod, ApplicationProject } from '../types/application';
import { Visibility } from '../types/component';

export function getApplicationLeak(
  application: string,
  branch?: string
): Promise<ApplicationPeriod[]> {
  return getJSON('/api/applications/show_leak', { application, branch }).then(
    r => r.leaks,
    throwGlobalError
  );
}

export function getApplicationDetails(application: string, branch?: string): Promise<Application> {
  return getJSON('/api/applications/show', { application, branch }).then(
    r => r.application,
    throwGlobalError
  );
}

export function addApplicationBranch(data: {
  application: string;
  branch: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/create_branch', data).catch(throwGlobalError);
}

export function updateApplicationBranch(data: {
  application: string;
  branch: string;
  name: string;
  project: string[];
  projectBranch: string[];
}) {
  return post('/api/applications/update_branch', data).catch(throwGlobalError);
}

export function deleteApplicationBranch(application: string, branch: string) {
  return post('/api/applications/delete_branch', { application, branch }).catch(throwGlobalError);
}

export function getApplicationProjects(data: {
  application: string;
  p?: number;
  ps?: number;
  q?: string;
  selected: string;
}): Promise<{ paging: T.Paging; projects: ApplicationProject[] }> {
  return getJSON('/api/applications/search_projects', data).catch(throwGlobalError);
}

export function addProjectToApplication(application: string, project: string) {
  return post('/api/applications/add_project', { application, project }).catch(throwGlobalError);
}

export function removeProjectFromApplication(application: string, project: string) {
  return post('/api/applications/remove_project', { application, project }).catch(throwGlobalError);
}

export function refreshApplication(key: string) {
  return post('/api/applications/refresh', { key }).catch(throwGlobalError);
}

export function createApplication(
  name: string,
  description: string,
  key: string | undefined,
  visibility: string
): Promise<{
  application: {
    description?: string;
    key: string;
    name: string;
    visibility: Visibility;
  };
}> {
  return postJSON('/api/applications/create', { description, key, name, visibility }).catch(
    throwGlobalError
  );
}

export function deleteApplication(application: string) {
  return post('/api/applications/delete', { application }).catch(throwGlobalError);
}

export function editApplication(application: string, name: string, description: string) {
  return post('/api/applications/update', { name, description, application }).catch(
    throwGlobalError
  );
}
