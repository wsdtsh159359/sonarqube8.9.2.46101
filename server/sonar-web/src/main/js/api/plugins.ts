/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { findLastIndex } from 'lodash';
import { getJSON, post } from 'sonar-ui-common/helpers/request';
import { isDefined } from 'sonar-ui-common/helpers/types';
import throwGlobalError from '../app/utils/throwGlobalError';
import {
  AvailablePlugin,
  InstalledPlugin,
  PendingPluginResult,
  PluginType,
  Update
} from '../types/plugins';

export function getAvailablePlugins(): Promise<{
  plugins: AvailablePlugin[];
  updateCenterRefresh: string;
}> {
  return getJSON('/api/plugins/available').catch(throwGlobalError);
}

export function getPendingPlugins(): Promise<PendingPluginResult> {
  return getJSON('/api/plugins/pending').catch(throwGlobalError);
}

function getLastUpdates(updates: undefined | Update[]): Update[] {
  if (!updates) {
    return [];
  }
  const lastUpdate = ['COMPATIBLE', 'REQUIRES_SYSTEM_UPGRADE', 'DEPS_REQUIRE_SYSTEM_UPGRADE'].map(
    status => {
      const index = findLastIndex(updates, update => update.status === status);
      return index > -1 ? updates[index] : undefined;
    }
  );
  return lastUpdate.filter(isDefined);
}

function addChangelog(update: Update, updates?: Update[]) {
  if (!updates) {
    return update;
  }
  const index = updates.indexOf(update);
  const previousUpdates = index > 0 ? updates.slice(0, index) : [];
  return { ...update, previousUpdates };
}

function getInstalledPluginApi(type = PluginType.External) {
  return getJSON('/api/plugins/installed', { f: 'category', type });
}

function getUpdatesPluginApi() {
  return getJSON('/api/plugins/updates');
}

export function getInstalledPlugins(
  type: PluginType = PluginType.External
): Promise<InstalledPlugin[]> {
  return getInstalledPluginApi(type).then(({ plugins }) => plugins, throwGlobalError);
}

export function getInstalledPluginsWithUpdates(): Promise<InstalledPlugin[]> {
  return Promise.all([getInstalledPluginApi(), getUpdatesPluginApi()])
    .then(([installed, updates]) =>
      installed.plugins.map((plugin: InstalledPlugin) => {
        const updatePlugin: InstalledPlugin = updates.plugins.find(
          (p: InstalledPlugin) => p.key === plugin.key
        );
        if (updatePlugin) {
          return {
            ...updatePlugin,
            ...plugin,
            updates: getLastUpdates(updatePlugin.updates).map(update =>
              addChangelog(update, updatePlugin.updates)
            )
          };
        }
        return plugin;
      })
    )
    .catch(throwGlobalError);
}

export function getPluginUpdates(): Promise<InstalledPlugin[]> {
  return Promise.all([getUpdatesPluginApi(), getInstalledPluginApi()])
    .then(([updates, installed]) =>
      updates.plugins.map((updatePlugin: InstalledPlugin) => {
        const updates = getLastUpdates(updatePlugin.updates).map(update =>
          addChangelog(update, updatePlugin.updates)
        );
        const plugin = installed.plugins.find((p: InstalledPlugin) => p.key === updatePlugin.key);
        if (plugin) {
          return {
            ...plugin,
            ...updatePlugin,
            updates
          };
        }
        return { ...updatePlugin, updates };
      })
    )
    .catch(throwGlobalError);
}

export function installPlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/install', data).catch(throwGlobalError);
}

export function uninstallPlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/uninstall', data).catch(throwGlobalError);
}

export function updatePlugin(data: { key: string }): Promise<void | Response> {
  return post('/api/plugins/update', data).catch(throwGlobalError);
}

export function cancelPendingPlugins(): Promise<void | Response> {
  return post('/api/plugins/cancel_all').catch(throwGlobalError);
}
