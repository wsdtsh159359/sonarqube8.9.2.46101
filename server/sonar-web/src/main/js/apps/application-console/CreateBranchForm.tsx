/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { some, without } from 'lodash';
import * as React from 'react';
import { ResetButtonLink, SubmitButton } from 'sonar-ui-common/components/controls/buttons';
import SimpleModal from 'sonar-ui-common/components/controls/SimpleModal';
import DeferredSpinner from 'sonar-ui-common/components/ui/DeferredSpinner';
import MandatoryFieldMarker from 'sonar-ui-common/components/ui/MandatoryFieldMarker';
import MandatoryFieldsExplanation from 'sonar-ui-common/components/ui/MandatoryFieldsExplanation';
import { translate } from 'sonar-ui-common/helpers/l10n';
import {
  addApplicationBranch,
  getApplicationDetails,
  updateApplicationBranch
} from '../../api/application';
import { Application, ApplicationProject } from '../../types/application';
import ProjectBranchRow from './ProjectBranchRow';
import { ApplicationBranch, SelectBranchOption } from './utils';

interface Props {
  application: Application;
  branch?: ApplicationBranch;
  enabledProjectsKey: string[];
  onClose: () => void;
  onCreate?: (branch: ApplicationBranch) => void;
  onUpdate?: (name: string) => void;
}

interface BranchesList {
  [name: string]: SelectBranchOption | null;
}

interface State {
  loading: boolean;
  name: string;
  projects: ApplicationProject[];
  selected: string[];
  selectedBranches: BranchesList;
}

const MAX_PROJECTS_HEIGHT = 220;
const PROJECT_HEIGHT = 22;
export default class CreateBranchForm extends React.PureComponent<Props, State> {
  mounted = false;
  node?: HTMLElement | null = null;
  currentSelect?: HTMLElement | null = null;

  state: State = {
    loading: false,
    name: '',
    projects: [],
    selected: [],
    selectedBranches: {}
  };

  componentDidMount() {
    this.mounted = true;
    const { application } = this.props;
    const branch = this.props.branch ? this.props.branch.name : undefined;
    this.setState({ loading: true });
    getApplicationDetails(application.key, branch).then(
      ({ projects }) => {
        if (this.mounted) {
          const enabledProjects = projects.filter(p =>
            this.props.enabledProjectsKey.includes(p.key)
          );
          const selected = enabledProjects.filter(p => p.selected).map(p => p.key);
          const selectedBranches: BranchesList = {};
          enabledProjects.forEach(p => {
            if (!p.enabled) {
              selectedBranches[p.key] = null;
            } else {
              selectedBranches[p.key] = {
                value: p.branch || '',
                label: p.branch || '',
                isMain: p.isMain || false
              };
            }
          });
          this.setState({
            name: branch || '',
            selected,
            loading: false,
            projects: enabledProjects,
            selectedBranches
          });
        }
      },
      () => {
        this.props.onClose();
      }
    );
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  stopLoading = () => {
    if (this.mounted) {
      this.setState({ loading: false });
    }
  };

  canSubmit = () => {
    const hasUnselectedBranches = some(this.state.selectedBranches, (branch, projectKey) => {
      return !branch && this.state.selected.includes(projectKey);
    });
    return (
      !this.state.loading &&
      this.state.name.length > 0 &&
      !hasUnselectedBranches &&
      this.state.selected.length > 0
    );
  };

  handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    this.setState({ name: event.currentTarget.value });
  };

  handleFormSubmit = async () => {
    const projectKeys = this.state.selected;

    const projectBranches = projectKeys.map(p => {
      const branch = this.state.selectedBranches[p];
      return !branch || branch.isMain ? '' : branch.value;
    });

    if (this.props.branch) {
      await updateApplicationBranch({
        application: this.props.application.key,
        branch: this.props.branch.name,
        name: this.state.name,
        project: projectKeys,
        projectBranch: projectBranches
      });
      if (this.props.onUpdate) {
        this.props.onUpdate(this.state.name);
      }
    } else {
      await addApplicationBranch({
        application: this.props.application.key,
        branch: this.state.name,
        project: projectKeys,
        projectBranch: projectBranches
      });
      if (this.props.onCreate) {
        this.props.onCreate({ name: this.state.name, isMain: false });
      }
    }
    this.props.onClose();
  };

  handleProjectCheck = (checked: boolean, key: string) => {
    this.setState(state => ({
      selected: checked ? [...state.selected, key] : without(state.selected, key)
    }));
  };

  handleBranchChange = (projectKey: string, branch: SelectBranchOption) => {
    this.setState(state => ({
      selectedBranches: { ...state.selectedBranches, [projectKey]: branch }
    }));
  };

  handleSelectorClose = () => {
    if (this.node) {
      this.node.classList.add('selector-hidden');
    }
  };

  handleSelectorDirection = (selectNode: HTMLElement, elementCount: number) => {
    if (this.node) {
      const modalTop = this.node.getBoundingClientRect().top;
      const modalHeight = this.node.offsetHeight;
      const maxSelectHeight = Math.min(MAX_PROJECTS_HEIGHT, (elementCount + 1) * PROJECT_HEIGHT);
      const selectBottom = selectNode.getBoundingClientRect().top + maxSelectHeight;
      if (selectBottom > modalTop + modalHeight) {
        this.node.classList.add('inverted-direction');
      } else {
        this.node.classList.remove('inverted-direction');
      }
      this.node.classList.remove('selector-hidden');
    }
  };

  renderProjectsList = () => {
    return (
      <>
        <strong className="spacer-left spacer-top">
          {translate('application_console.branches.configuration')}
        </strong>
        <p className="spacer-top big-spacer-bottom spacer-left spacer-right">
          {translate('application_console.branches.create.help')}
        </p>
        <table className="data zebra">
          <thead>
            <tr>
              <th className="thin" />
              <th className="thin">{translate('project')}</th>
              <th>{translate('branch')}</th>
            </tr>
          </thead>
          <tbody>
            {this.state.projects.map(project => (
              <ProjectBranchRow
                checked={this.state.selected.includes(project.key)}
                key={project.key}
                onChange={this.handleBranchChange}
                onCheck={this.handleProjectCheck}
                onClose={this.handleSelectorClose}
                onOpen={this.handleSelectorDirection}
                project={project}
              />
            ))}
          </tbody>
        </table>
      </>
    );
  };

  render() {
    const isUpdating = this.props.branch !== undefined;
    const header = translate('application_console.branches', isUpdating ? 'update' : 'create');
    return (
      <SimpleModal
        header={header}
        onClose={this.props.onClose}
        onSubmit={this.handleFormSubmit}
        size="medium">
        {({ onCloseClick, onFormSubmit, submitting }) => (
          <form className="views-form" onSubmit={onFormSubmit}>
            <div className="modal-head">
              <h2>{header}</h2>
            </div>

            <div
              className="modal-body modal-container selector-hidden"
              ref={node => (this.node = node)}>
              {this.state.loading ? (
                <div className="text-center big-spacer-top big-spacer-bottom">
                  <i className="spinner spacer-right" />
                </div>
              ) : (
                <>
                  <MandatoryFieldsExplanation className="modal-field" />
                  <div className="modal-field">
                    <label htmlFor="view-edit-name">
                      {translate('name')}
                      <MandatoryFieldMarker />
                    </label>
                    <input
                      autoFocus={true}
                      className="input-super-large"
                      maxLength={250}
                      name="name"
                      onChange={this.handleInputChange}
                      size={50}
                      type="text"
                      value={this.state.name}
                    />
                  </div>
                  {this.renderProjectsList()}
                </>
              )}
            </div>

            <div className="modal-foot">
              <DeferredSpinner className="spacer-right" loading={submitting} />
              <SubmitButton disabled={submitting || !this.canSubmit()}>
                {translate(
                  'application_console.branches',
                  isUpdating ? 'update' : 'create',
                  'verb'
                )}
              </SubmitButton>
              <ResetButtonLink onClick={onCloseClick}>
                {translate('application_console.branches.cancel')}
              </ResetButtonLink>
            </div>
          </form>
        )}
      </SimpleModal>
    );
  }
}
