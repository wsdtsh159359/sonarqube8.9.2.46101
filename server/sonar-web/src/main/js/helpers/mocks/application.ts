/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { Application, ApplicationPeriod, ApplicationProject } from '../../types/application';
import { Visibility } from '../../types/component';
import { mockBranch } from './branch-like';

export function mockApplication(overrides: Partial<Application> = {}): Application {
  return {
    branches: [mockBranch()],
    key: 'foo',
    name: 'Foo',
    projects: [mockApplicationProject()],
    visibility: Visibility.Private,
    ...overrides
  };
}

export function mockApplicationPeriod(
  overrides: Partial<ApplicationPeriod> = {}
): ApplicationPeriod {
  return {
    date: '2017-10-01',
    project: 'foo',
    projectName: 'Foo',
    ...overrides
  };
}

export function mockApplicationProject(
  overrides: Partial<ApplicationProject> = {}
): ApplicationProject {
  return {
    branch: 'master',
    isMain: true,
    key: 'bar',
    name: 'Bar',
    ...overrides
  };
}
