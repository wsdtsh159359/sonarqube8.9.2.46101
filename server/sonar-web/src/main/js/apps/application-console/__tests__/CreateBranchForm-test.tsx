/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { waitAndUpdate } from 'sonar-ui-common/helpers/testUtils';
import { getApplicationDetails } from '../../../api/application';
import { mockApplication, mockApplicationProject } from '../../../helpers/mocks/application';
import { Application } from '../../../types/application';
import CreateBranchForm from '../CreateBranchForm';

jest.mock('../../../api/application', () => ({
  getApplicationDetails: jest.fn().mockResolvedValue({}),
  addApplicationBranch: jest.fn(),
  updateApplicationBranch: jest.fn()
}));

it('Should handle submit correctly', async () => {
  const app = mockApplication();
  (getApplicationDetails as jest.Mock<Promise<Application>>).mockResolvedValueOnce(app);
  const handleClose = jest.fn();
  const handleCeate = jest.fn();
  const handleUpdate = jest.fn();
  let wrapper = shallowRender({ application: app, onClose: handleClose, onCreate: handleCeate });
  wrapper.instance().handleFormSubmit();
  await waitAndUpdate(wrapper);
  expect(handleClose).toHaveBeenCalled();
  expect(handleCeate).toHaveBeenCalled();

  (getApplicationDetails as jest.Mock<Promise<Application>>).mockResolvedValueOnce(app);
  wrapper = shallowRender({
    application: app,
    branch: { isMain: false, name: 'foo' },
    onUpdate: handleUpdate
  });
  wrapper.instance().handleFormSubmit();
  await waitAndUpdate(wrapper);
  expect(handleUpdate).toHaveBeenCalled();
});

it('Should render correctly', async () => {
  const app = mockApplication({
    projects: [
      mockApplicationProject({ key: '1', enabled: true }),
      mockApplicationProject({ key: '2', enabled: true }),
      mockApplicationProject({ key: '3', enabled: false }),
      mockApplicationProject({ enabled: false })
    ]
  });
  (getApplicationDetails as jest.Mock<Promise<Application>>).mockResolvedValueOnce(app);
  const wrapper = shallowRender({ application: app, enabledProjectsKey: ['1', '3'] });
  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();
});

it('Should close when no response', async () => {
  const app = mockApplication();
  (getApplicationDetails as jest.Mock<Promise<Application>>).mockRejectedValueOnce(app);
  const handleClose = jest.fn();
  const wrapper = shallowRender({ application: app, onClose: handleClose });
  await waitAndUpdate(wrapper);
  expect(handleClose).toHaveBeenCalled();
});

it('Should update loading flag', () => {
  const wrapper = shallowRender();
  wrapper.setState({ loading: true });
  wrapper.instance().stopLoading();
  expect(wrapper.state().loading).toBe(false);
});

it('Should update on input event', () => {
  const wrapper = shallowRender();
  wrapper.setState({ name: '' });
  wrapper
    .instance()
    .handleInputChange(({ currentTarget: { value: 'bar' } } as any) as React.ChangeEvent<
      HTMLInputElement
    >);
  expect(wrapper.state().name).toBe('bar');
});

it('Should tell if it can submit correctly', () => {
  const wrapper = shallowRender();
  wrapper.setState({ loading: true });
  expect(wrapper.instance().canSubmit()).toBe(false);
  wrapper.setState({ loading: false, name: '' });
  expect(wrapper.instance().canSubmit()).toBe(false);
  wrapper.setState({
    loading: false,
    name: 'ok',
    selectedBranches: { foo: null },
    selected: ['foo']
  });
  expect(wrapper.instance().canSubmit()).toBe(false);
  wrapper.setState({
    loading: false,
    name: 'ok',
    selectedBranches: { foo: { label: 'foo', isMain: true, value: 'foo' } },
    selected: ['foo']
  });
  expect(wrapper.instance().canSubmit()).toBe(true);
});

function shallowRender(props: Partial<CreateBranchForm['props']> = {}) {
  return shallow<CreateBranchForm>(
    <CreateBranchForm
      application={mockApplication()}
      enabledProjectsKey={[]}
      onClose={jest.fn()}
      onCreate={jest.fn()}
      onUpdate={jest.fn()}
      {...props}
    />
  );
}
