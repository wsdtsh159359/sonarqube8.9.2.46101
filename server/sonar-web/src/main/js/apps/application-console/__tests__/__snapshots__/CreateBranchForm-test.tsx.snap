// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render correctly 1`] = `
<SimpleModal
  header="application_console.branches.create"
  onClose={[MockFunction]}
  onSubmit={[Function]}
  size="medium"
>
  <Component />
</SimpleModal>
`;
